File: app/src/main/java/com/example/sellerkeyboard/ImageKeyboard.java
package com.example.sellerkeyboard;

import android.content.ClipDescription;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.inputmethodservice.InputMethodService;
import android.net.Uri;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputConnection;
import android.widget.Button;
import android.widget.GridLayout;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RawRes;
import androidx.core.view.inputmethod.EditorInfoCompat;
import androidx.core.view.inputmethod.InputConnectionCompat;
import androidx.core.view.inputmethod.InputContentInfoCompat;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.storage.FileDownloadTask;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;


public class ImageKeyboard extends InputMethodService {

    private static final String TAG = "ImageKeyboard";
    private static final String AUTHORITY = "com.example.sellerkeyboard.fileprovider";
    private static final String MIME_TYPE_GIF = "image/gif";
    private static final String MIME_TYPE_PNG = "image/png";

    ArrayList<Button> buttons = new ArrayList<>();
    private List<Snippet> snippetItemList = new ArrayList<>();


    private boolean isCommitContentSupported(
            @Nullable EditorInfo editorInfo, @NonNull String mimeType) {
        if (editorInfo == null) {
            return false;
        }

        final InputConnection ic = getCurrentInputConnection();
        if (ic == null) {
            return false;
        }

        if (validatePackageName(editorInfo)) {
            return false;
        }

        final String[] supportedMimeTypes = EditorInfoCompat.getContentMimeTypes(editorInfo);
        for (String supportedMimeType : supportedMimeTypes) {
            if (ClipDescription.compareMimeTypes(mimeType, supportedMimeType)) {
                return true;
            }
        }
        return false;
    }

    private static int getFlag() {
//        return InputConnectionCompat.INPUT_CONTENT_GRANT_READ_URI_PERMISSION; //
        return 1;
    }

    private boolean validatePackageName(@Nullable EditorInfo editorInfo) {
        if (editorInfo == null) {
            return true;
        }
        final String packageName = editorInfo.packageName;
        return packageName == null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
    }

    //Create a button to insert GIF, PNG, WebP
    @Override
    public View onCreateInputView() {
        snippetItemList = new ArrayList<>(); // Initialize snippetItemList
        fetchDataFromSQLite();
        buttons.clear(); // Clear existing buttons
        for (Snippet snippetItem : snippetItemList) {
            Button imageButton = new Button(this);
            imageButton.setText(snippetItem.getTitle());
            imageButton.setOnClickListener(view -> {
                new Thread(() -> {
                    try {
                        Random random = new Random();
                        InputConnection inputConnection = getCurrentInputConnection();

                        Thread.sleep(random.nextInt(100) + 400);
                        if (inputConnection != null) {
                            inputConnection.commitText(snippetItem.getContent(), 1);
                        } else {
                            Log.d("Onclick", "inputConnection is null");
                        }

                        Thread.sleep(500);
                        assert inputConnection != null;
                        inputConnection.performEditorAction(EditorInfo.IME_ACTION_SEND);
                        Thread.sleep(500);

                        // Send imageUrls
                        ImageKeyboard.this.doCommitContent(
                                "imageUrls",
                                snippetItem.getImageUrl()
                        );

                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }).start();
            });
            buttons.add(imageButton);
        }

        //Grid Layout
        final GridLayout layout = new GridLayout(this);
        layout.setColumnCount(3); // Set the desired number of columns
        for (Button button : buttons) {
            layout.addView(button);
        }
        return layout;
    }

    private void fetchDataFromSQLite() {
        SnippetDbHelper dbHelper = new SnippetDbHelper(this);
        snippetItemList = dbHelper.getAllSnippets();
    }



    private void doCommitContent(String description, String imageUrls) {
        //get Image from local storage
        Glide.with(this)
                .load(imageUrls)
                .into(
                        new CustomTarget<Drawable>() {
                            @Override
                            public void onResourceReady(@NonNull Drawable resource, @Nullable Transition<? super Drawable> transition) {
                                Bitmap bitmap = ((BitmapDrawable) resource).getBitmap();
                                String path = MediaStore.Images.Media.insertImage(getContentResolver(), bitmap, "Image Description", null);
                                Uri uri = Uri.parse(path);
                                final int flag = getFlag();
                                final InputContentInfoCompat inputContentInfoCompat = new InputContentInfoCompat(
                                        uri,
                                        new ClipDescription(description, new String[]{MIME_TYPE_PNG}),
                                        null
                                );
                                InputConnectionCompat.commitContent(
                                        getCurrentInputConnection(), getCurrentInputEditorInfo(), inputContentInfoCompat,
                                        flag, null);
                            }

                            @Override
                            public void onLoadCleared(@Nullable Drawable placeholder) {
                            }
                        }
                );

        final EditorInfo editorInfo = getCurrentInputEditorInfo();

        // Validate packageName again just in case.
        if (validatePackageName(editorInfo)) {
            return;
        }
    }


    @Override
    public boolean onEvaluateFullscreenMode() {
        // In full-screen mode the inserted content is likely to be hidden by the IME. Hence in this
        // sample we simply disable full-screen mode.
        return false;
    }

    //
    @Override
    public void onStartInputView(EditorInfo info, boolean restarting) {
        //send EditorInfo.contentMimeTypes
        Log.d(TAG, "onStartInputView: " + Arrays.toString(info.contentMimeTypes));


        for (Button button : buttons) {
            button.setEnabled(true);
        }
    }
}

File: app/src/main/kotlin/com/terasumi/sellerkeyboard/ImageKeyboard.kt
package com.terasumi.sellerkeyboard;

import android.content.ClipDescription;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.inputmethodservice.InputMethodService;
import android.net.Uri;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputConnection;
import android.widget.Button;
import android.widget.GridLayout;
import android.widget.ImageButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.FileProvider;
import androidx.core.content.res.ResourcesCompat;
import androidx.core.view.inputmethod.InputConnectionCompat;
import androidx.core.view.inputmethod.InputContentInfoCompat;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.google.android.voiceime.VoiceRecognitionTrigger;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;


public class ImageKeyboard extends InputMethodService {

    private static final String TAG = "ImageKeyboard";
    private static final String MIME_TYPE_PNG = "image/png";
//    File tempFile = null;


    ArrayList<Button> buttons = new ArrayList<>();
    private List<Snippet> snippetItemList = new ArrayList<>();
    private VoiceRecognitionTrigger mVoiceRecognitionTrigger;
    //    Button mButton;
    Button openGoogleVoiceButton;
    GridLayout mLayout;


    private static int getFlag() {
        return InputConnectionCompat.INPUT_CONTENT_GRANT_READ_URI_PERMISSION; //
    }

    @Override
    public void onCreate() {
        super.onCreate();
        // Initialize VoiceRecognitionTrigger
        Log.i(TAG, "#onCreate");

        // Create the voice recognition trigger, and register the listener.
        // The trigger has to unregistered, when the IME is destroyed.
        mVoiceRecognitionTrigger = new VoiceRecognitionTrigger(this);
        mVoiceRecognitionTrigger.register(new VoiceRecognitionTrigger.Listener() {

            @Override
            public void onVoiceImeEnabledStatusChange() {
                // The call back is done on the main thread.
                updateVoiceImeStatus();
            }
        });
    }

    /**
     * Update the microphone icon to reflect the status of the voice recognition.
     */
    private void updateVoiceImeStatus() {
        Log.i(TAG, "#updateVoiceImeStatus");

        if (openGoogleVoiceButton == null) {
            Log.i(TAG, "openGoogleVoiceButton is null");
            return;
        }

        if (mVoiceRecognitionTrigger.isInstalled()) {
            openGoogleVoiceButton.setVisibility(View.VISIBLE);
            if (mVoiceRecognitionTrigger.isEnabled()) {
                // Voice recognition is installed and enabled.
                Log.i(TAG, "Voice recognition is enabled.");
                openGoogleVoiceButton.setEnabled(true);
            } else {
                // Voice recognition is installed, but it is not enabled (no network).
                // The microphone icon is displayed greyed-out.
                Log.i(TAG, "Voice recognition is disabled.");
                openGoogleVoiceButton.setEnabled(false);
            }
        } else {
            // Voice recognition is not installed, and the microphone icon is not displayed.
            Log.i(TAG, "Voice recognition is not installed.");
            openGoogleVoiceButton.setVisibility(View.GONE);
        }
        mLayout.invalidate();
    }

    /**
     * Returns the language of the IME. The langauge is used in voice recognition to match the
     * current language of the IME.
     */
    private String getImeLanguage() {
        return "vi-VN";
    }

    //Create a button to insert GIF, PNG, WebP
    @Override
    public View onCreateInputView() {
        snippetItemList = new ArrayList<>(); // Initialize snippetItemList
        fetchDataFromSQLite();
        buttons.clear(); // Clear existing buttons
        for (Snippet snippetItem : snippetItemList) {
            Button imageButton = getButton(snippetItem);
            buttons.add(imageButton);
        }

        //contraint layout
        View rootView = getLayoutInflater().inflate(R.layout.keyboard, null);
        mLayout = rootView.findViewById(R.id.buttonGrid);
        //config layout
//        openGoogleVoiceButton = new ImageButton(this);
        openGoogleVoiceButton = new Button(this);
        openGoogleVoiceButton.setText("Voice");
        Drawable voiceDrawable = ResourcesCompat.getDrawable(getResources(), R.drawable.voice, null);
        voiceDrawable.setBounds(0, 0, 50, 50); // Set the desired width and height
        openGoogleVoiceButton.setCompoundDrawables(voiceDrawable, null, null, null);        openGoogleVoiceButton.setHeight(3);

        openGoogleVoiceButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mVoiceRecognitionTrigger.startVoiceRecognition(getImeLanguage());
            }
        });

        buttons.add(0, openGoogleVoiceButton);

        // Get the screen width
        int screenWidth = getResources().getDisplayMetrics().widthPixels - 6; // 6 is the padding of the layout
        // Set the number of columns

        for (Button button : buttons) {
            button.setGravity(Gravity.CENTER);
            //buton = screen width / 4
            button.setWidth(screenWidth / 4);
            mLayout.addView(button);
        }
        return mLayout;
    }

    private @NonNull Button getButton(Snippet snippetItem) {
        Button imageButton = new Button(this);
        imageButton.setText(snippetItem.getTitle());
        imageButton.setOnClickListener(view -> new Thread(() -> {
            try {
                Random random = new Random();
                InputConnection inputConnection = getCurrentInputConnection();

                Thread.sleep(random.nextInt(100) + 400);
                if (inputConnection != null) {
                    inputConnection.commitText(snippetItem.getContent(), 1);
                } else {
                    Log.d("Onclick", "inputConnection is null");
                }

                Thread.sleep(500);
                assert inputConnection != null;
                inputConnection.performEditorAction(EditorInfo.IME_ACTION_SEND);
//                    inputConnection.performEditorAction(EditorInfo.IME_ACTION_DONE);
                Thread.sleep(500);

                // Send imageUrls
                ImageKeyboard.this.doCommitContent(snippetItem.getImageUrl());

            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }).start());
        return imageButton;
    }

    private void fetchDataFromSQLite() {
        SnippetDbHelper dbHelper = new SnippetDbHelper(this);
        snippetItemList = dbHelper.getAllSnippets();
    }


    private void doCommitContent(String imageUrl) {
        Glide.with(this).load(imageUrl).into(new CustomTarget<Drawable>() {
            @Override
            public void onResourceReady(@NonNull Drawable resource, @Nullable Transition<? super Drawable> transition) {
                Bitmap bitmap = ((BitmapDrawable) resource).getBitmap();
                try {
                    File tempFile = File.createTempFile("temp_image", ".jpg", getCacheDir());
                    try (FileOutputStream outputStream = new FileOutputStream(tempFile)) {
                        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, outputStream);
                    }
                    Uri uri = FileProvider.getUriForFile(ImageKeyboard.this, "com.terasumi.sellerkeyboard.fileprovider", tempFile);
                    InputContentInfoCompat inputContentInfo = new InputContentInfoCompat(uri, new ClipDescription("imageUrls", new String[]{"image/jpeg"}), null);
                    InputConnectionCompat.commitContent(getCurrentInputConnection(), getCurrentInputEditorInfo(), inputContentInfo, getFlag(), null);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }

            @Override
            public void onLoadCleared(@Nullable Drawable placeholder) {
            }
        });
    }


    @Override
    public boolean onEvaluateFullscreenMode() {
        // In full-screen mode the inserted content is likely to be hidden by the IME. Hence in this
        // sample we simply disable full-screen mode.
        return false;
    }

    //
    @Override
    public void onStartInputView(EditorInfo info, boolean restarting) {
        if (buttons != null) {
            for (Button button : buttons) {
                button.setEnabled(true);
            }
        }
    }

    @Override
    public void onDestroy() {
        Log.i(TAG, "#onDestroy");
        super.onDestroy();

        if (mVoiceRecognitionTrigger != null) {
            mVoiceRecognitionTrigger.unregister(this);
        }
    }
}